src/extension/vectortiles/src/main/java/org/geotools/renderer/lite/VectorMapRenderUtils.java

```diff
@@ -443,4 +445,42 @@ private static boolean isFeatureTypeStyleActive(FeatureType ftype, FeatureTypeSt
                                 || FeatureTypes.isDecendedFrom(
                                         ftype, null, fts.getFeatureTypeName())));
     }
+
+    /**
+     * This function executes transformations on the vector features e.g. point stacker.
+     *
+     * @param layer the layer to process
+     * @param paintArea the area that is painted on
+     * @param mapScale the current map scale
+     * @param schema the schema of the features
+     * @param features the feature collection typically points
+     * @return the transformed feature collection
+     * @throws IOException
+     */
+    public static FeatureCollection<?, ?> transformIfNecessary(
+            Layer layer,
+            Rectangle paintArea,
+            double mapScale,
+            FeatureType schema,
+            FeatureCollection<?, ?> features)
+            throws IOException {
+
+        List<LiteFeatureTypeStyle> featureTypeStyles =
+                VectorMapRenderUtils.getFeatureStyles(layer, paintArea, mapScale, schema);
+        if (!featureTypeStyles.isEmpty()) {
+            for (LiteFeatureTypeStyle featureTypeStyle : featureTypeStyles) {
+                try {
+                    if (featureTypeStyle.transformation != null) {
+                        Expression transform = featureTypeStyle.transformation;
+                        features = (FeatureCollection<?, ?>) transform.evaluate(features);
+                    }
+                } catch (Exception ex) {
+                    if (LOGGER.isLoggable(Level.WARNING)) {
+                        LOGGER.log(Level.WARNING, "Error while appliying transformation", ex);
+                    }
```

Comment:
"appliying" is misspelt - should be "applying"

Also, the error message is a little vague; it might be useful to include exactly what transformation it is trying to apply.

Reply:
@tbarsballe when getting into detail what problem happens another null pointer exception might happen. From my perspective the exception itself should contain detailed information what went wrong.

Reply:
Thats what null checks are for. Depending on what the exception is, the error message may not be obvious. 

In particular, if there is a ClassCastException on line 475 because the Transformation did not return a FeatureCollection, then its not going to be obvious that the actual error is that the style is using an incompatible rendering transform.